File: 2020/python/day4.py
from std import *
import re
import itertools

props = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid']

def isValid(pp):
    if len(pp) < len(props) or len(pp) > len(props) + 1:
        return False
    for p in props:
        if p not in pp:
            return False
    return True

pps = []
pp = {}
for line in lines("../input/input4.txt"):
    if not line:
        pps.append(pp)
        pp = {}
        continue 
    for p in [x.split(":") for x in line.split(" ")]:        
        pp[p[0]] = p[1]        
if pp:        
    pps.append(pp)


pps = list(filter(isValid, pps))
print("A:", len(pps))

count = 0
for pp in pps:

    byr = pp['byr']
    if len(byr) != 4 or byr[0] == '0' or int(byr) < 1920 or int(byr) > 2002:
        continue

    iyr = pp['iyr']
    if len(iyr) != 4 or iyr[0] == '0' or int(iyr) < 2010 or int(iyr) > 2020:
        continue

    eyr = pp['eyr']
    if len(eyr) != 4 or eyr[0] == '0' or int(eyr) < 2020 or int(eyr) > 2030:
        continue

    height = pp['hgt']
    if not re.match("[1-9][0-9]+(in|cm)", height):
        continue

    h = int(height[:-2])
    u = height[-2:]    

    if u == 'cm' and (h < 150 or h > 193):
        continue
    if u == 'in' and (h < 59 or h > 76):
        continue

    if not re.match("#[0-9a-f]{6}", pp['hcl']):
        continue

    if pp['ecl'] not in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']: 
        continue        

    if not match(r"[0-9]{9}", pp['pid']):
        continue
    count += 1

print("B:", count)

File: 2020/haskell/day4.hs
import Data.Map (Map, fromList)
import Data.Maybe
import Data.List.Split (splitOn)
import Data.List (sort)
import Text.Regex
import Debug.Trace

inRange (low, high) value = low <= value && value <= high

group :: [String] -> [String] -> [[String]]
group [] a = [a]
group (x:xs) a = case length x of
        0 -> a : group xs []
        _ -> group xs (a ++ [x])

hasAll :: [String] -> [String] -> Bool
hasAll xs ys = foldl (\a x -> elem x ys && a) True xs

parse :: [String] -> [(String,String)]
parse xs = map ((\(a:b) -> (a, head b)) . splitOn ":") $ concatMap (splitOn " ") xs

hasAllProperties :: [(String, String)] -> Bool
hasAllProperties xs = hasAll ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"] (map fst xs)

isValid :: (String, String) -> Bool
isValid ("byr", value) = inRange (1920, 2002) year where year = read value :: Int
isValid ("iyr", value) = inRange (2010, 2020) year where year = read value :: Int
isValid ("eyr", value) = inRange (2020, 2030) year where year = read value :: Int
isValid ("hgt", value) = case matchRegex (mkRegex "([0-9]+)(in|cm)") value of
                                Just [height, unit] -> inRange interval (read height :: Int) 
                                        where interval = if unit == "cm" then (150, 193) else (59, 76)
                                _ -> False
isValid ("hcl", value) = isJust $ matchRegex (mkRegex "^#[0-9a-f]{6}$") value 
isValid ("ecl", value) = value `elem` ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
isValid ("pid", value) = isJust $ matchRegex (mkRegex "^[0-9]{9}$") value 
isValid ("cid", _) = True
isValid _ = False

showP _ [] = []
showP p (x:xs) = if p == fst x then [x] else showP p xs

main = do 
        c <- readFile "../input/input4.txt"
        let field = lines c
        let passports = map parse (group field [])
        let valid = filter hasAllProperties passports
        putStrLn ("A: " ++ show (length valid))
        putStrLn ("B: " ++ show (length (filter (all isValid) valid)))
